File: backend/src/llm/anthropicProvider.ts
import Anthropic from '@anthropic-ai/sdk';
import type { LlmProvider, LineGuess, GoalGuess } from './provider.js';

const LINE_PROMPT = (line: string, unresolved: string[]) => `You are a gym-log parser. Given this logged line: "${line}"
The unrecognized tokens are: ${unresolved.join(', ')}.
Respond ONLY with JSON: {"exerciseName": string, "equipment": string|null, "weightKg": number|null, "reps": number|null, "sets": number|null}`;

const GOAL_PROMPT = (text: string) => `You are a fitness goal classifier. Given this goal description: "${text}"
Respond ONLY with JSON: {"type": "HYPERTROPHY"|"STRENGTH"|"ENDURANCE"|"CUSTOM", "muscles": string[]} where muscles are from
["GLUTES","QUADS","HAMSTRINGS","CHEST","BACK","SHOULDERS","ARMS","CORE","CALVES"].`;

export class AnthropicProvider implements LlmProvider {
  private client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });
  private model = 'claude-haiku-4-5-20251001';

  private async ask(prompt: string): Promise<string> {
    const msg = await this.client.messages.create({
      model: this.model,
      max_tokens: 256,
      messages: [{ role: 'user', content: prompt }],
    });
    const block = msg.content[0];
    if (block.type !== 'text') throw new Error('unexpected non-text response from Anthropic');
    return block.text;
  }

  async resolveLine(line: string, unresolvedTokens: string[]): Promise<LineGuess> {
    const raw = await this.ask(LINE_PROMPT(line, unresolvedTokens));
    return JSON.parse(raw) as LineGuess;
  }

  async resolveGoal(text: string): Promise<GoalGuess> {
    const raw = await this.ask(GOAL_PROMPT(text));
    return JSON.parse(raw) as GoalGuess;
  }
}

File: backend/internal/llm/anthropic_provider.go
package llm

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/anthropics/anthropic-sdk-go"
	"github.com/anthropics/anthropic-sdk-go/option"
)

type AnthropicProvider struct {
	client anthropic.Client
	model  anthropic.Model
}

func NewAnthropicProvider(apiKey string) *AnthropicProvider {
	return &AnthropicProvider{
		client: anthropic.NewClient(option.WithAPIKey(apiKey)),
		model:  anthropic.ModelClaudeHaiku4_5,
	}
}

func (p *AnthropicProvider) ask(ctx context.Context, prompt string) (string, error) {
	msg, err := p.client.Messages.New(ctx, anthropic.MessageNewParams{
		Model:     p.model,
		MaxTokens: 256,
		Messages: []anthropic.MessageParam{
			anthropic.NewUserMessage(anthropic.NewTextBlock(prompt)),
		},
	})
	if err != nil {
		return "", err
	}
	if len(msg.Content) == 0 || msg.Content[0].Type != "text" {
		return "", fmt.Errorf("anthropic: unexpected non-text response")
	}
	return msg.Content[0].Text, nil
}

func (p *AnthropicProvider) ResolveLine(ctx context.Context, line string, unresolved []string) (LineGuess, error) {
	raw, err := p.ask(ctx, linePrompt(line, unresolved))
	if err != nil {
		return LineGuess{}, err
	}
	var guess LineGuess
	if err := json.Unmarshal([]byte(raw), &guess); err != nil {
		return LineGuess{}, fmt.Errorf("anthropic: parse line guess: %w", err)
	}
	return guess, nil
}

func (p *AnthropicProvider) ResolveGoal(ctx context.Context, text string) (GoalGuess, error) {
	raw, err := p.ask(ctx, goalPrompt(text))
	if err != nil {
		return GoalGuess{}, err
	}
	var guess GoalGuess
	if err := json.Unmarshal([]byte(raw), &guess); err != nil {
		return GoalGuess{}, fmt.Errorf("anthropic: parse goal guess: %w", err)
	}
	return guess, nil
}
